include/aspect/particle/property/cpo.h

```diff
@@ -0,0 +1,361 @@
+           */
+          double
+          advect_Crank_Nicolson(std::vector<double> &volume_fractions,
+                                std::vector<Tensor<2,3> > &a_cosine_matrices,
+                                const std::pair<std::vector<double>, std::vector<Tensor<2,3> > > &derivatives,
+                                std::vector<double> &previous_volume_fraction_derivatives,
+                                std::vector<Tensor<2,3> > &previous_a_cosine_matrices_derivatives,
+                                const double strain_rate_second_invariant,
+                                const double dt) const;
+
+
+          /**
+           * A interface function which delegates the computation the derivatives
+           * to specialized functions.
+           */
+          std::pair<std::vector<double>, std::vector<Tensor<2,3> > >
+          compute_derivatives(const std::vector<double> &volume_fractions,
+                              const std::vector<Tensor<2,3> > &a_cosine_matrices,
+                              const SymmetricTensor<2,3> &strain_rate_nondimensional,
+                              const Tensor<2,3> &velocity_gradient_tensor_nondimensional,
+                              const double volume_fraction_mineral,
+                              const std::array<double,4> &ref_resolved_shear_stress) const;
+
+          /**
+           * Computes the update for the property rotating passively with the particle.
+           */
+          std::pair<std::vector<double>, std::vector<Tensor<2,3> > >
+          compute_derivatives_spin_tensor(const std::vector<double> &volume_fractions,
+                                          const std::vector<Tensor<2,3> > &a_cosine_matrices,
+                                          const SymmetricTensor<2,3> &strain_rate_nondimensional,
+                                          const Tensor<2,3> &velocity_gradient_tensor_nondimensional,
+                                          const double volume_fraction_mineral,
+                                          const std::array<double,4> &ref_resolved_shear_stress) const;
+
+
+          /**
+           * Declare the parameters this class takes through input files.
+           */
+          static
+          void
+          declare_parameters (ParameterHandler &prm);
+
+          /**
+           * Read the parameters this class declares from the parameter file.
+           */
+          void
+          parse_parameters (ParameterHandler &prm) override;
+
+          /**
+           * Return the number of grains per particle
+           */
+          static
+          unsigned int
+          get_number_of_grains();
+
+          /**
+           * Return the number of minerals per particle
+           */
+          static
+          unsigned int
+          get_number_of_minerals();
+
+        private:
+          // helpful shortcuts
+          double rad_to_degree = 180.0/M_PI;
```

Comment:
I have not looked into the code, but I can not imagine having these shortcuts as member variables is significantly better than just defining them as `const double` in the function where you need them. Also we do not usually use `M_PI`, we use `numbers::PI` instead

Reply:
I just removed them for now.